src/vision/status_quo/alan_finds_database_drops_hard.md

```diff
@@ -0,0 +1,104 @@
+  
+```rust
+use sqlx::Connection;
+
+#[async_std::main]
+async fn main() -> Result<(), sqlx::Error> {
+    // Create a connection
+
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?;
+
+    assert_eq!(row.0, 150);
+
+    Ok(())
+}
+```
+
+Things seem to be working fairly well but sometimes when he refreshes the page he encounters a panic with the message "Cannot open a new connecton: connection is already open". He is flummoxed.
+
+
+## Searching for the Solution
+
+
+Alan tries to figure out what happened from the logs, but the only information he sees is that a new connection has been received. Alan turns to the documentation for the `sqlx` crate to see if there are flags that might enable extra instrumentation but he can't find any [sqlx::Connection::close](https://docs.rs/sqlx/0.5.1/sqlx/trait.Connection.html#required-methods).
+
+* He does find the [`close` method] which mentions "This method is not required for safe and consistent operation. However, it is recommended to call it instead of letting a connection drop as the database backend will be faster at cleaning up resources."
+* He adds a call to `close` into his code and it helps some but he is still able to reproduce the problem if he refreshes often enough. 
+* He adds a log statement right before calling `close` to see if it is working:
+
+```rust
+use sqlx::Connection;
+
+#[async_std::main]
+async fn do_the_thing() -> Result<(), sqlx::Error> {
+    // Create a connection
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?; // <----- if this await is cancelled, doesn't help
+
+    assert_eq!(row.0, 150);
+    
+    // he adds this:
+    log!("closing the connection");
+    conn.close();
+
+    Ok(())
+}
+```
+  
+* He observes that in the cases where he has the problem the log statement never executes.
+* He asks Barbara for help and she explains how `await` can be canceled and it will the destructors for things that are in scope.
+* He reads the [source for the SqliteConnection destructor](https://github.com/launchbadge/sqlx/blob/0ed524d65c2a3ee2e2a6706910b85bf2bb72115f/sqlx-core/src/pool/connection.rs#L70-L74) and finds that destructor spawns a task to actually close the connection.
+* He realizes there is a race condition:
+    * the task may not have actually closed the connection before `do_the_thing` is called a second time
+* He gives up and gets frustrated
+
+Next, Alan seeks verification and validation of his understanding of the source code from the `sqlx` forum. His advisor states that Rust doesn't have a way to execute async operations in a destructor.
```

Comment:
```suggestion
Next, Alan seeks verification and validation of his understanding of the source code from the `sqlx` forum. Someone on the forum explains why the destructor launches a fresh task: Rust doesn't have a way to execute async operations in a destructor.
```

but I wonder if it would be better to remove this and move it into the story above. It seems more likely he would ask on the forum and be directed to the source code than to go digging and find it directly. Doesn't matter too much though.